guardrails/validator_service.py

```diff
@@ -0,0 +1,232 @@
+                f"key {validator_setup.key} has value {value}."
+            )
+        return value, metadata
+
+    async def validate_dependents(
+        self, value, metadata, validator_setup, validation_logs
+    ):
+        async def process_child(child_setup):
+            child_value = value[child_setup.key]
+            child_validation_logs = FieldValidationLogs()
+            validation_logs.children[child_setup.key] = child_validation_logs
+            new_child_value, new_metadata = await self.async_validate(
+                child_value,
+                metadata,
+                child_setup,
+                child_validation_logs,
+            )
+            return child_setup.key, new_child_value, new_metadata
+
+        tasks = [process_child(child_setup) for child_setup in validator_setup.children]
+        results = await asyncio.gather(*tasks)
+
+        for key, child_value, child_metadata in results:
+            value[key] = child_value
+            # TODO address conflicting metadata entries
+            metadata = {**metadata, **child_metadata}
+
+        return value, metadata
+
+    async def async_validate(
+        self,
+        value: Any,
+        metadata: dict,
+        validator_setup: FieldValidation,
+        validation_logs: FieldValidationLogs,
+    ) -> Tuple[Any, dict]:
+        # Validate children first
+        if validator_setup.children:
+            await self.validate_dependents(
+                value, metadata, validator_setup, validation_logs
+            )
+
+        # Validate the field
+        value, metadata = await self.run_validators(
+            validation_logs, validator_setup, value, metadata
+        )
+
+        return value, metadata
+
+    def validate(
+        self,
+        value: Any,
+        metadata: dict,
+        validator_setup: FieldValidation,
+        validation_logs: FieldValidationLogs,
+    ) -> Tuple[Any, dict]:
+        # Run validate_async in an async loop
+        loop = asyncio.get_event_loop()
+        if loop.is_running():
+            raise RuntimeError(
+                "Async event loop found, please call `validate_async` instead."
+            )
+        value, metadata = loop.run_until_complete(
+            self.async_validate(
+                value,
+                metadata,
+                validator_setup,
+                validation_logs,
+            )
+        )
+        return value, metadata
+
+
+def validate(
+    value: Any,
+    metadata: dict,
+    validator_setup: FieldValidation,
+    validation_logs: FieldValidationLogs,
+):
+    loop = asyncio.get_event_loop()
+    if loop.is_running():
+        logger.warning(
+            "Async event loop found, but guard was invoked synchronously."
+            "For validator parallelization, please call `validate_async` instead."
+        )
+        validator_service = SequentialValidatorService()
+    else:
+        validator_service = AsyncValidatorService()
```

Comment:
Is the goal here to try to force the validators to run in parallel whenever possible?  If not, why switch between sync and async when sync was inherently requested via this being called rather than async_validate?

Reply:
> Is the goal here to try to force the validators to run in parallel whenever possible?

Yes